libfreerdp/core/childsession.c

```diff
@@ -0,0 +1,332 @@
+		return -1;
+
+	return written;
+}
+
+static int transport_bio_named_read(BIO* bio, char* buf, int size)
+{
+	WINPR_BIO_NAMED* ptr = (WINPR_BIO_NAMED*)BIO_get_data(bio);
+
+	if (!buf)
+		return 0;
+
+	BIO_clear_flags(bio, BIO_FLAGS_READ);
+
+	DWORD readBytes;
+	BOOL ret = ReadFile(ptr->hFile, buf, size, &readBytes, NULL);
+	WLog_VRB(TAG, "transport_bio_named_read(%d)=%d read=%d", size, ret, readBytes);
+	if (!ret)
+	{
+		if (GetLastError() == ERROR_NO_DATA)
+			BIO_set_flags(bio, (BIO_FLAGS_SHOULD_RETRY | BIO_FLAGS_READ));
+		return -1;
+	}
+
+	if (readBytes == 0)
+	{
+		BIO_set_flags(bio, (BIO_FLAGS_READ | BIO_FLAGS_SHOULD_RETRY));
+		return 0;
+	}
+
+	return readBytes;
+}
+
+static int transport_bio_named_puts(BIO* bio, const char* str)
+{
+	return transport_bio_named_write(bio, str, strlen(str));
+}
+
+static int transport_bio_named_gets(BIO* bio, char* str, int size)
+{
+	return transport_bio_named_write(bio, str, size);
+}
+
+static long transport_bio_named_ctrl(BIO* bio, int cmd, long arg1, void* arg2)
+{
+	int status = -1;
+	WINPR_BIO_NAMED* ptr = (WINPR_BIO_NAMED*)BIO_get_data(bio);
+
+	switch (cmd)
+	{
+		case BIO_C_SET_SOCKET:
+		case BIO_C_GET_SOCKET:
+			return -1;
+		case BIO_C_GET_EVENT:
+			if (!BIO_get_init(bio) || !arg2)
+				return 0;
+
+			*((HANDLE*)arg2) = ptr->hFile;
+			return 1;
+		case BIO_C_SET_HANDLE:
+			BIO_set_init(bio, 1);
+			if (!BIO_get_init(bio) || !arg2)
+				return 0;
+
+			ptr->hFile = (HANDLE)arg2;
+			return 1;
+		case BIO_C_SET_NONBLOCK:
+		{
+			return 1;
+		}
+		case BIO_C_WAIT_READ:
+		{
+			int timeout = (int)arg1;
+			return 1;
+		}
+		break;
+		case BIO_C_WAIT_WRITE:
+		{
+			int timeout = (int)arg1;
+			return 1;
+		}
+		break;
+		default:
+			break;
+	}
+
+	switch (cmd)
+	{
+		case BIO_CTRL_GET_CLOSE:
+			status = BIO_get_shutdown(bio);
+			break;
+
+		case BIO_CTRL_SET_CLOSE:
+			BIO_set_shutdown(bio, (int)arg1);
+			status = 1;
+			break;
+
+		case BIO_CTRL_DUP:
+			status = 1;
+			break;
+
+		case BIO_CTRL_FLUSH:
+			status = 1;
+			break;
+
+		default:
+			status = 0;
+			break;
+	}
+
+	return status;
+}
+
+static int transport_bio_named_uninit(BIO* bio)
+{
+	WINPR_BIO_NAMED* ptr = (WINPR_BIO_NAMED*)BIO_get_data(bio);
+
+	if (ptr && ptr->hFile)
+	{
+		CloseHandle(ptr->hFile);
+		ptr->hFile = NULL;
+	}
+
+	BIO_set_init(bio, 0);
+	BIO_set_flags(bio, 0);
+	return 1;
+}
+
+static int transport_bio_named_new(BIO* bio)
+{
+	BIO_set_flags(bio, BIO_FLAGS_SHOULD_RETRY);
+
+	WINPR_BIO_NAMED* ptr = (WINPR_BIO_NAMED*)calloc(1, sizeof(WINPR_BIO_NAMED));
+	if (!ptr)
+		return 0;
+
+	BIO_set_data(bio, ptr);
+	BIO_set_flags(bio, BIO_FLAGS_SHOULD_RETRY);
+	return 1;
+}
+
+static int transport_bio_named_free(BIO* bio)
+{
+	WINPR_BIO_NAMED* ptr;
+
+	if (!bio)
+		return 0;
+
+	transport_bio_named_uninit(bio);
+	ptr = (WINPR_BIO_NAMED*)BIO_get_data(bio);
+
+	if (ptr)
+	{
+		BIO_set_data(bio, NULL);
+		free(ptr);
+	}
+
+	return 1;
+}
+
+BIO_METHOD* BIO_s_namedpipe(void)
+{
+	static BIO_METHOD* bio_methods = NULL;
```

Comment:
sure it is a good idea to make this a single instance without thread sync?

Reply:
Well do that for all the other BIOs...

Reply:
https://www.openssl.org/docs/manmaster/man3/BIO_new.html also agrees

Reply:
I also agree, but we should probably do that for all the instances in a commit of its own.